crates/solidity/outputs/cargo/crate/src/generated/bindings/generated/built_ins/0.8.24.sol

```diff
@@ -0,0 +1,105 @@
+// This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+contract $BuiltIns$ {
+  function addmod(uint x, uint y, uint k) public returns (uint);
+  function assert(bool condition) public;
+  function blockhash(uint blockNumber) public returns (bytes32);
+  function blobhash(uint index) public returns (bytes32);
+  function ecrecover(bytes32 hash, uint8 v, bytes32 r, bytes32 s) public returns (address);
+  function gasleft() public returns (uint256);
+  function keccak256(bytes memory) public returns (bytes32);
+  function mulmod(uint x, uint y, uint k) public returns (uint);
+  function require(bool condition) public;
+  function require(bool condition, string memory message) public;
+  function revert() public;
+  function revert(string memory reason) public;
+  function ripemd160(bytes memory) public returns (bytes20);
+  function selfdestruct(address payable recipient) public;
+  function sha256(bytes memory) public returns (bytes32);
+  struct $abiType {
+    function(bytes memory, $args) returns ($args) decode;
+    function($args) returns (bytes memory) encode;
+    function(function(), $args) returns (bytes memory) encodeCall;
+    function($args) returns (bytes memory) encodePacked;
+    function(bytes4 selector, $args) returns (bytes memory) encodeWithSelector;
+    function(string memory, $args) returns (bytes memory) encodeWithSignature;
+  }
+  struct $address {
+    uint256 balance;
+    bytes code;
+    bytes32 codehash;
+    function(bytes memory) returns (bool, bytes memory) call;
+    function(bytes memory) returns (bool, bytes memory) delegatecall;
+    function(uint256) returns (bool) send;
+    function(bytes memory) returns (bool, bytes memory) staticcall;
+    function(uint256) transfer;
+  }
+  struct $array {
+    uint length;
```

Comment:
moving from the other thread:

I see `$element` referenced in 5 places, but with no definition. What would that type resolve to in the generated `built_ins.sol`?

Reply:
context: I'm trying to understand how generics are simulated here, and so for `type()` expressions.

Reply:
I wonder, with the custom rule file for built-ins, (should we) avoid exposing such internal concepts to the user? IIUC, there is no Solidity-compatible way to represent this in syntax (or any generic arguments for that matter). Thinking out loud, for both our and end-user DX, should we just bind this in the rules, without needing the (illegal) type here?

Reply:
There are two things here, which are separate:

1. Given a reference, what is the location of what we deem as the definition.
2. What's the semantic binding of that reference, eg. what's its _type_; this we need when the reference node is not what we're trying to resolve, but it's in the middle of a resolution stack (eg. the operand in a `MemberAccessExpression`).

For 2. we can certainly encode the semantics in the rules when it's not possible to express that in Solidity, and hence it's not possible to build a portion of the graph to convey the binding semantics. What happens in this case is that we hard-code in the rules exactly what we want the graph to look like.

But we still need to address point 1. Ie. where will we point the user to when they ask for the definition of one of these references? `this` and `super` we can point to the contract or the parents as you suggested. But what to do in other cases?

Reply:
For (1), I [replied here](https://github.com/NomicFoundation/slang/pull/1125#discussion_r1841392782). Maybe `this`/`super` don't need to be references, as they are now keywords not identifiers?

For (2), I think it is a case-by-case decision. I trust your judgement here if you think the existing solution works. My main concern is finding a way to expose a semantically accurate (as much as possible) `built-ins.sol` to the user.

Reply:
I think it is a case-by-case decision. I trust your judgement here if you think the existing solution works. My main concern is finding a way to expose a semantically accurate (as much as possible) `built-ins.sol` to the user.